### PR TITLE
feat: plugin context binding

### PR DESCRIPTION
**Pull request by underfin**
### Description

The rollup has tree kind of plugin context, here has some description.

- `PluginContext`, used at `buildStart/load/resolveId...` hooks
- `TransformPluginContext`, used at `transform` hook
- `MminimalPluginContext`, used at `options` hook

The derived relationship is `MminimalPluginContext` -> `PluginContext` -> `TransformPluginContext`, the `MminimalPluginContext` does not need implemented at the rust side because `options` hook is at js side. So the rust side only implement `PluginContext`/`TransformPluginContext` and binding.

The `PluginContext` is created for use in one the plugin's different hooks, because it's `cache` filed is isolate, ref https://github.com/rollup/rollup/blob/master/src/utils/PluginDriver.ts#L108.

The `TransformPluginContext` is derived from  `PluginContext`, it extends some methods and also overwrite some methods, you can see https://github.com/rollup/rollup/blob/master/src/utils/transform.ts#L108.

### Implement

The rust side is also implements `PluginContext`  and `TransformPluginContext`, using `PluginContext` to create `TransformPluginContext` and custom some methods. 

Here move `PluginContext` to immutable because its `cache` need to make sure it thread-safe.

### Test Plan

Not need at now.

---

### CODE REVIEW

File: crates/rolldown/src/bundler/plugin_driver/mod.rs
Comment (milesj) [inline crates/rolldown/src/bundler/plugin_driver/mod.rs:16]: Is context different per plugin? Shouldn't context be the same across all plugins?

Comment (underfin) [reply crates/rolldown/src/bundler/plugin_driver/mod.rs:16]: Yeah. The plugin context is different, you can see here https://github.com/rollup/rollup/blob/master/src/utils/PluginDriver.ts#L108.

**Comment by underfin** [comment]
* **#400** <a href="https://app.graphite.dev/github/pr/rolldown-rs/rolldown/400?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#399** <a href="https://app.graphite.dev/github/pr/rolldown-rs/rolldown/399?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @underfin and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Comment by hyf0** [comment]
I'm going to on hold thid PR until we have real-world usages to verify if the changes are meaningful. Especially, some code-deleting are not understandable to me.

**Review by milesj — APPROVED** [review]
@underfin Changes here look good, is there any other pieces missing before this lands?

**Comment by underfin** [comment]
> @underfin Changes here look good, is there any other pieces missing before this lands?

Need review from @hyf0 .

**Comment by hyf0** [comment]
I just looked around all PRs related to plugin. They are ok under the current plugin system code. But there are some problems in the current plugin system code and these PRs are just patch the bad code and add features on top of it. That's bad for having a solid ground.

So I will remove the bad code and come up with new one. So I'm gonna keep holding up these PRs before the new plugin system code.

**Pull request closed without merging**
